[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Nahin-Imtiaz/CSC448--Artificial-Intelligence/blob/master/Assignment_4/Assignment_4.ipynb)

In [1]:
# Imports section
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely, not locally)
iris = datasets.load_iris()
df = pd.DataFrame(iris.data, columns = iris.feature_names)

# Output the first 15 rows of the data
df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                 5.1               3.5                1.4               0.2
1                 4.9               3.0                1.4               0.2
2                 4.7               3.2                1.3               0.2
3                 4.6               3.1                1.5               0.2
4                 5.0               3.6                1.4               0.2
5                 5.4               3.9                1.7               0.4
6                 4.6               3.4                1.4               0.3
7                 5.0               3.4                1.5               0.2
8                 4.4               2.9                1.4               0.2
9                 4.9               3.1                1.5               0.1
10                5.4               3.7                1.5               0.2
11                4.8               3.4                1.6               0.2
12                4.8               3.0                1.4               0.1
13                4.3               3.0                1.1               0.1
14                5.8               4.0                1.2               0.2

In [3]:
# Display a summary of the table information (number of datapoints, etc.)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


In [4]:
df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)  
count        150.000000  
mean           1.199333  
std            0.762238  
min            0.100000  
25%            0.300000  
50%            1.300000  
75%            1.800000  
max            2.500000

# About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

There are four features. They are sepal length, sepal width, petal length, and petal width. There are 150 samples in the dataset. There are no null values. There’s also a column called target. The label for this assignment is the target. The label is split into 3 subclasses with numeric values - 0 for Iris-Setosa, 1 for Iris-Versicolour, and 2 for Iris-Virginica.

# Part 2: Split the dataset into train and test

In [5]:
# Take the dataset and split it into our features (X) and label (y)
features_x = iris.data
label_y = iris.target

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(features_x, label_y, train_size = 0.9, test_size = 0.1, random_state=42)

# Part 3: Logistic Regression

In [6]:
# i. Use sklearn to train a LogisticRegression model on the training set
Logistic = LogisticRegression()
Logistic.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
pred = Logistic.predict_proba(X_test[:1])
print("Predictions: ", pred)

Predictions:  [[0.00345824 0.83719406 0.1593477 ]]


In [7]:
# iii. Report on the score for Logistic regression model, what does the score measure?
print(Logistic.score(X_test, y_test))

1.0


The score of our model is 1.0 which means that the model will always predict the right class of iris.

In [8]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
print('Coefficients:',Logistic.coef_)

Coefficients: [[-0.42671215  0.97262703 -2.44459812 -1.03197327]
 [ 0.51311126 -0.22370085 -0.21512542 -0.85154854]
 [-0.08639911 -0.74892618  2.65972354  1.88352181]]


In [9]:
print('Intercepts:',Logistic.intercept_)

Intercepts: [  9.50102574   1.91208599 -11.41311173]


# Part 4: Support Vector Machine

In [10]:
# i. Use sklearn to train a Support Vector Classifier on the training set
SVM = svm.SVC(probability=True)
SVM.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
pred = SVM.predict_proba(X_test[:1])
print("Predictions: ", pred)

Predictions:  [[0.00686332 0.90372833 0.08940835]]


In [11]:
# iii. Report on the score for the SVM, what does the score measure?
SVM.score(X_test, y_test)

1.0

The score of our model is 1.0 which means that the model will always predict the right class of iris.

# Part 5: Neural Network

In [12]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
NN = MLPClassifier(hidden_layer_sizes=(12, 6), max_iter=1000)
NN.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
pred = NN.predict_proba(X_test[:1])
print("Predictions: ", pred)

Predictions:  [[0.00150417 0.94914885 0.04934697]]


C:\Users\Nahin Imtiaz\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:582: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


In [13]:
# iii. Report on the score for the Neural Network, what does the score measure?
NN.score(X_test, y_test)

1.0

The score of our model is 1.0 which means that the model will always predict the right class of iris.

In [14]:
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)

# Quasi-Newton Method
NN1 = MLPClassifier(solver = 'lbfgs', hidden_layer_sizes=(12, 6), max_iter=1000)
NN1.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
pred = NN1.predict_proba(X_test[:1])
print("Predictions: ", pred)

Predictions:  [[9.30638485e-142 1.00000000e+000 3.63995066e-114]]


In [15]:
NN1.score(X_test, y_test)

1.0

In [16]:
# Stochastic Gradient Descent
NN2 = MLPClassifier(solver = 'sgd', hidden_layer_sizes=(12, 6), max_iter=1000)
NN2.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
pred = NN2.predict_proba(X_test[:1])
print("Predictions: ", pred)

Predictions:  [[0.00152993 0.87427124 0.12419883]]


In [17]:
NN2.score(X_test, y_test)

1.0

I kept all the configurations the same except for the solvers. For the first configuration, the solver is adam, which is the default, and the score I got is 1.0. The next two solvers are lbfgs and sgd and the score for both of them is 1.0.

# Part 6: K-Nearest Neighbors

In [26]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
KNN = KNeighborsClassifier(n_neighbors=5)
KNN.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
KNN.predict_proba(X_test[:1])

array([[0., 1., 0.]])

In [27]:
# iii. Report on the score for kNN, what does the score measure?
KNN.score(X_test, y_test)

1.0

The score of our model is 1.0 which means that the model will always predict the right class of iris.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

In this assignment, we have to use 4 different Classifiers to analyze which type gives the best score. The 4 types of Classifiers are Logistic Regression, Support Vector Machine, Neural Network, and K-Nearest Neighbors. To my surprise, all 4 models give a score of 1. It could be due to the dataset. Another surprising thing is that K-Nearest Neighbors is giving 100% probability for each class. It could be due to overfitting the model or having duplicates in train and test splits. It is expected to have high accuracies.